client/extensions/woocommerce/app/products/product-form-variations-card.js

```diff
@@ -12,69 +12,121 @@ import ProductVariationTypesForm from './product-variation-types-form';
 import ProductFormVariationsTable from './product-form-variations-table';
 import FormToggle from 'components/forms/form-toggle';
 
-const ProductFormVariationsCard = ( {
-	product,
-	variations,
-	editProduct,
-	translate,
-	editProductAttribute,
-	editProductVariation
-} ) => {
-	// TODO When toggling between types, don't destory a user's previous edits. Save to component state and restore on toggle.
-	const handleToggle = () => {
-		if ( 'variable' !== product.type ) {
-			editProduct( product, { type: 'variable' } );
+class ProductFormVariationsCard extends Component {
+
+	state = {
+		simpleProduct: [],
+		variationAttributes: [],
+	};
+
+	static propTypes = {
+		product: PropTypes.shape( {
+			type: PropTypes.string.isRequired,
+			name: PropTypes.string,
+			attributes: PropTypes.array,
+		} ),
+		variations: PropTypes.array,
+		editProduct: PropTypes.func.isRequired,
+		editProductAttribute: PropTypes.func.isRequired,
+		editProductVariation: PropTypes.func.isRequired,
+	};
+
+	/*
+	 * When switching between product types, we clean up the Redux state to drop fields or variation attributes
+	 * that are no longer valid, so we can keep API calls clean. This is done in setVariable and setSimple.
+	 * To preserve a users edits, we store these changes in component state, and restore them if they toggle back.
+	 * A user shouldn't loose work just because they locally toggled a card.
+	 */
+	handleToggle = () => {
+		const simpleFields = [ 'dimensions', 'weight', 'regular_price', 'manage_stock', 'stock_quantity', 'backorders' ];
```

Comment:
This seems like it would be more appropriate as a class constant instead of being passed to these two class members.